File: packages/backend/app/models/form_entry.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import String, Integer, Text, DateTime, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class FormEntry(Base):
    __tablename__ = "form_entries"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    module: Mapped[str] = mapped_column(String(100), index=True)
    date: Mapped[str] = mapped_column(String(10), index=True)
    user_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("usuarios.id")
    )
    device_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("devices.id")
    )
    version: Mapped[int] = mapped_column(Integer, default=1)
    data_json: Mapped[str] = mapped_column(Text)
    checksum: Mapped[str | None] = mapped_column(String(64))
    status: Mapped[str] = mapped_column(String(20))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

File: packages/frontend/lib/domain/entities/form_entry.dart
class FormEntry {
  final String id;
  final String module;
  final String date;
  final String userId;
  final String deviceId;
  final int version;
  final Map<String, dynamic> data;
  final String status;

  const FormEntry({
    required this.id,
    required this.module,
    required this.date,
    this.userId = '',
    this.deviceId = '',
    this.version = 1,
    this.data = const {},
    this.status = 'saved',
  });

  factory FormEntry.fromJson(Map<String, dynamic> json) {
    return FormEntry(
      id: json['id'] as String,
      module: json['module'] as String,
      date: json['date'] as String,
      userId: json['user_id'] as String? ?? '',
      deviceId: json['device_id'] as String? ?? '',
      version: json['version'] as int? ?? 1,
      data: json['data'] as Map<String, dynamic>? ?? {},
      status: json['status'] as String? ?? 'saved',
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'module': module,
        'date': date,
        'user_id': userId,
        'device_id': deviceId,
        'version': version,
        'data': data,
        'status': status,
      };
}
